### ISSUE TITLE
debug: compile-time debug flags for Emily.Backend (opt-in assertions)

### ISSUE THREAD

**Issue by ausimian** [ausimian/emily#32]
## Background

During DistilBERT-QA conformance testing we hit an intermittent `:nan`
score on M4. Root cause: `Bumblebee.Text.question_answering` was paired
with a vocab-30522 tokenizer against a tiny-random model whose
embedding table has only 1124 rows. The tokenizer emitted token ids
past the table bound, and `Emily.Backend.gather/4` silently returned
uninitialised memory for the out-of-range rows — usually benign
floats, occasionally a `NaN` that propagated through softmax.

`Nx.BinaryBackend` raises on OOB gather because the check is
effectively free on CPU. `Emily.Backend` (like EXLA / JAX-GPU /
Torch-CUDA) dispatches to a GPU kernel (`mx::take`, `mx::gather`) that
does not bounds-check. This is the GPU-backend norm, not an Emily bug
per se — but it hides a real class of caller error.

## Proposal

Add a compile-time opt-in debug-assertion facility to `Emily.Backend`
using `Application.compile_env/2`. When the flag is `false` (the
default), the assertion branch is dead-code-eliminated by the Elixir
compiler — zero runtime cost, zero bytecode. Consumers flip the flag
in `config/test.exs` (or CI) and Mix recompiles the affected modules
automatically.

Sketch:

```elixir
@bounds_check_gather Application.compile_env(:emily, :bounds_check_gather, false)

def gather(out, input, indices, opts) do
  if @bounds_check_gather, do: assert_in_bounds!(input.shape, indices, opts[:axes])
  ...
end
```

## Candidate flags to consider pre-release

Non-exhaustive — the milestone should scope a coherent set, not
ship every one:

- `:bounds_check_gather` — assert indices in range for `gather` /
  `take` (the case above).
- `:bounds_check_scatter` — same, but for writes; scatters also hit
  embedding-adjacent paths (position embeddings, ngram tables).
- `:detect_nan_inf` — post-op NaN/Inf scan on the hot training ops
  (matmul / softmax / layer-norm) to pinpoint where numerics break,
  instead of chasing it from a downstream loss.
- `:assert_same_backend` — detect tensors that silently moved between
  `Emily.Backend` and `Nx.BinaryBackend` mid-graph (common footgun
  when a helper returns a host tensor).
- `:strict_dtype` — raise instead of silently promoting on mixed-
  dtype ops; useful during the M16 mixed-precision work.

## Notes

- Each assertion will typically cost a GPU sync (breaks lazy-graph
  fusion), which is why they must be opt-in.
- The cost is only paid in builds with the flag on — so "default off"
  is critical and must be verified (inspect `.beam` disassembly or
  rely on the `Application.compile_env/2` warning when
  compile-vs-runtime values diverge).
- This is scoped as a **pre-release milestone** rather than a single
  PR — we want a thought-through set of flags, consistent naming
  (`:debug_*`?), and a documented story in `README.md` / `CHANGELOG.md`
  for downstream consumers, not a piecemeal rollout.
